config/emacs/config.org
#+TITLE: GNU Emacs Config
#+AUTHOR: rho
#+DESCRIPTION: my personal Emacs config
#+STARTUP: overview
#+STARTUP: latexpreview

~C-h C-h~  'help-for-help'
~C-h f~    'describe-function'
~C-h v~    'describe-variable'
~C-h P~    'describe-package'
=C-h k=

C-c ;    to disable a section
:tangle  to disable a block

-----

(local-variable-p 'completions-header-format)

M-\
and
M-<SPC>

* interface
#+begin_src emacs-lisp
  ;; to avoid the following message:
  ;; 'Symbolic link to Git-controlled source file; follow link? (yes or no)'
  (setq-default vc-follow-symlinks t)

  ;; make ESC quit prompts
  ;;(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src

#+begin_src emacs-lisp
  ;; Emacs will not show its default welcome/startup screen when it launches
  (setq inhibit-startup-message t)
#+end_src

** theme/font
#+begin_src emacs-lisp
  (set-face-attribute 'default nil
                      :foreground "#ffffff"
                      :background "#000000")

  (set-face-attribute 'default nil :font "Hack" :height 120)
#+end_src

** menus/icons/buttons
#+begin_src emacs-lisp
  (scroll-bar-mode -1)  ; disables visible scrollbar
  (tool-bar-mode -1)    ; remove the row of clickable icons like 'open/save'
  (menu-bar-mode -1)    ; remove the top bar 'File/Edit/Options'
  ;;(global-tab-line-mode t) ; always shows active buffers
  ;;(tab-bar-mode t)         ; always shows active tabs
#+end_src

** fringes
#+begin_src emacs-lisp
  ;; fringes:
  ;; empty strips on the left and right of the text area, it shows:
  ;; - line continuation indicators
  ;; - error or warning markers (from tools like LSP or Flycheck)
  ;; - breakpoints or diff markers
  (set-fringe-mode 0) ; default is 10
#+end_src

** misc
#+begin_src emacs-lisp
  ;; hides the mouse pointer when it’s not being used.
  (setq make-pointer-invisible t) ; emacs must be running in graphical mode
#+end_src


** cursor
#+begin_src emacs-lisp
  ;;(setq-default cursor-type t) ; the default block cursor
  ;;(setq-default cursor-type 'bar)
  (blink-cursor-mode 0)

  ;;	The cursor normally appears in non-selected windows as a
  ;; non-blinking hollow box.  (For a bar cursor, it instead appears
  ;; as a thinner bar.)  To turn off cursors in non-selected windows,
  ;; change the variable
  ;; ‘cursor-in-non-selected-windows’ to ‘nil’.
  (setq-default cursor-in-non-selected-windows nil)

  ;;(setq x-setrech-cursor t)
#+end_src

** highlighting
#+begin_src emacs-lisp
   (set-face-attribute 'highlight nil
     :background "#333333")
  ;; inverse-video t
#+end_src

** line
#+begin_src emacs-lisp
  ;; highlights the current line
  ;;(global-hl-line-mode 'toggle)
  (global-hl-line-mode)

  (setq-default display-line-numbers-width 2)

  (global-display-line-numbers-mode 1)
  (setq display-line-numbers-type 'visual)
  (setq display-line-numbers-current-absolute nil)


  ;;    In a narrowed buffer (*note Narrowing::) lines are normally numbered
  ;; starting at the beginning of the narrowing.  However, if you customize
  ;; the variable ‘display-line-numbers-widen’ to a non-‘nil’ value, line
  ;; numbers will disregard any narrowing and will start at the first
  ;; character of the buffer.


  ;;    The line numbers are displayed in a special face ‘line-number’.  The
  ;; current line number is displayed in a different face,
  ;; ‘line-number-current-line’, so you can make the current line’s number
  ;; have a distinct appearance, which will help locating the line showing
  ;; point.  Additional faces ‘line-number-major-tick’ and
  ;; ‘line-number-minor-tick’ can be used to highlight the line numbers of
  ;; lines which are a multiple of certain numbers.  Customize
  ;; ‘display-line-numbers-major-tick’ and ‘display-line-numbers-minor-tick’
  ;; respectively to set those numbers.

  (set-face-attribute 'line-number nil
    :foreground "#777777"
    :background "#000000") ; ':extend t' ?

  (set-face-attribute 'line-number-current-line nil
    :foreground "#777777"
    :background "#000000") ; ':extend t' ?

  ;; NOTE: I've used 'C-h o' on 'line-mode' to see its description
  ;; NOTE: 'C-h o' on 'set-face-attribute'/'line-number'
        ;; :inverse-video 'set-face-attribute' description

;;  (setq display-line-numbers-major-tick 10)
;;  (set-face-attribute 'line-number-major-tick nil
;;    :foreground "#777777"
;;    :background "#000000")

;;  (setq display-line-numbers-minor-tick 3)
;;  (set-face-attribute 'line-number-minor-tick nil
;;    :foreground "#777777"
;;    :background "#000000")
#+end_src

** mode-line
'C-h o' on 'mode-line'
#+begin_src emacs-lisp
  (set-face-attribute 'mode-line-active nil
    :foreground "#ffff00"
    :background "#222222"
    :box nil)

  (set-face-attribute 'mode-line-inactive nil
    :foreground "#777777"
    :background "#222222"
    :box nil)
#+end_src

*** column
#+begin_src emacs-lisp
  ;; to always show the column number by default
  (column-number-mode)
#+end_src

#+begin_src emacs-lisp
  ;; column count starts from 1 instead of 0
  (setq column-number-indicator-zero-based nil)

  ;; C-h v 'column-number-indicator-zero-based'
  ;; This variable is obsolete since 28.1; use
  ;; ‘mode-line-position-column-format’ instead.
#+end_src

*** size
#+begin_src emacs-lisp
  (size-indication-mode 1)
#+end_src

*** which-function
#+begin_src emacs-lisp
  (require 'which-func)
  (set-face-attribute 'which-func nil
  		    :foreground "#ff00ff")
#+end_src

** minibuffer
#+begin_src emacs-lisp
  (set-face-attribute 'minibuffer-prompt nil
    :foreground "#00ff00"
    :background "#000000")

;; minibuffer-default-prompt-format
#+end_src

#+begin_src emacs-lisp
  (setq completion-show-help nil)
  ;;(setq completions-header-format nil)
  ;; default value:
  ;; #("%s possible completions:
  ;; " 0 25 (face shadow))

  ;;(setq completions-format 'horizontal)
  (setq completions-format 'one-column)
  (setq completions-max-height 20)

  (set-face-attribute 'completions-highlight nil
    :foreground "#ff00ff"
    :background "#000000"
    :box nil)
#+end_src

** messages
#+begin_src emacs-lisp
  ;; when Emacs needs to alert you (like an invalid keypress or reaching the
  ;; top/bottom of a buffer), it flashes the screen or frame briefly
  (setq visible-bell t)

  ;; show partially typed keyboard shortcuts in the minibuffer after a very
  ;; short pause (0.1s)
  (setq echo-keystrokes 0.1)

  (setq suggest-key-bindings 1)
#+end_src
(local-variable-p 'suggest-key-bindings)

** help
C-N C-h i

#+begin_src emacs-lisp
  (which-key-mode 1)
#+end_src
	M-x shortdoc

** region
#+begin_src emacs-lisp
  (set-face-attribute 'region nil
    :background "#550055")
#+end_src

NOTE: if a region is active, M-% will work only on that region


** speedbar
#+begin_src emacs-lisp
  ;; just hook to disable visual lines when using speedbar
  ;; M-x speedbar
  (add-hook 'speedbar-mode-hook
            (lambda ()
              (display-line-numbers-mode 0)))
#+end_src


* text
;;(setq-default major-mode 'text-mode)
;;(setq-default truncate-lines t) ; change this variable

** filling
#+begin_src emacs-lisp
  ;; NOTE: to see the value of 'fill-column'
  ;; fill-column C-x C-e    (its default value is 70)
  (setq-default fill-column 70)

  (add-hook 'text-mode-hook 'turn-on-auto-fill)
  ;; NOTE: 'auto-fill-mode' automatically inserts real
  ;; line breaks when you type past 'fill-column'

  ;;(display-fill-column-indicator-mode 1)
  (global-display-fill-column-indicator-mode 1)
#+end_src
*** about text mode
   Text mode among other actions, sets the syntax table to work
conveniently for writers.  In Text mode, Emacs considers an apostrophe
as part of a word like a letter; but Emacs does not consider a period
or a space as part of a word.  Thus, ‘M-f’ moves you over ‘it's’.  On
the other hand, in C mode, ‘M-f’ stops just after the ‘t’ of ‘it's’.

** line wrapping
#+begin_src emacs-lisp
  ;; to wrap lines visually at the window edge;
  ;; follow screen lines, not logical lines
  (global-visual-line-mode)
#+end_src


** mark
   Other commands always operate on the region, and have no default
behavior.  Such commands usually have the word ‘region’ in their names,
like ‘C-w’ (‘kill-region’) and ‘C-x C-u’ (‘upcase-region’).  If the mark
is inactive, they operate on the “inactive region”—that is, on the text
between point and the position at which the mark was last set (*note
Mark Ring::).  To disable this behavior, change the variable
‘mark-even-if-inactive’ to ‘nil’.  Then these commands will instead
signal an error if the mark is inactive.
#+begin_src emacs-lisp
  ;;(local-variable-p 'mark-even-if-inactive)
  (setq mark-even-if-inactive nil)
#+end_src

-----

mark-ring
global-mark-ring

C-<SPC>
C-u C-<SPC>
C-x C-<SPC>

CRIAR UMA FUNÇÃO PARA DELETAR UM MARK DO MARK RING
(defun my-delete-current-mark ()
  (interactive)
  (when mark-ring
    (setq mark-ring (cdr mark-ring))
    (message "Deleted most recent mark")))

-----

#+begin_src emacs-lisp
  ;; C-x C-x to check where the mark is (exchanges the positions of
  ;; the point and the mark)

  ;; While Transient Mark mode is off, you can activate it temporarily
  ;; using ‘C-<SPC> C-<SPC>’ or ‘C-u C-x C-x’

  (transient-mark-mode 1) ; global minor mode
  ;; [...] One reason you may want to use them is that some commands
  ;; operate on the entire buffer instead of the region when Transient
  ;; Mark mode is off.  Enabling Transient Mark mode momentarily gives
  ;; you a way to use these commands on the region.
#+end_src

** kill
* code

[[* filling]]: fill-column 80 ; global-display-fill-column-indicator-mode
** trailing whitespace
#+begin_src emacs-lisp
  (setq-default show-trailing-whitespace t)
  ;;(setq-default indicate-empty-lines t) ; emacs must be running in graphical mode
#+end_src

** tabs/indenting
(local-variable-p 'indent-tabs-mode) maybe I should just use setq
#+begin_src emacs-lisp
  (setq-default tab-width 8)

  (setq-default indent-tabs-mode t) ; tabs for indentation
  ;;(setq-default indent-tabs-mode nil) ; spaces for indentation

  (setq-default tab-always-indent t)
  ;; 't': hitting TAB always just indents the current line
  ;; 'nil':  hitting TAB indents the current line if point is at the left margin
  ;; or in the line’s indentation, otherwise it inserts a "real" TAB character
  ;; 'complete': TAB first tries to indent the current line, and if the line
  ;; was already indented, then try to complete the thing at point
  	;;(setq tab-always-indent 'complete)
#+end_src
*** notes
C-q <TAB>    to insert a =tab character=
note: it ALWAYS inserts a tab, regardless of the value of 'indent-tabs-mode'

C-M-\        'ident-region'
M-x tabify
M-x untabify

C-x <TAB>    (note: I think a region must be active)
M-i

M-^
C-M-o

-----

note: 'eletric-indent-mode' is enabled by default


** cc mode
3 getting started
#+begin_src emacs-lisp
  (setq-default c-basic-offset 8)

  (setq-default c-default-style '((java-mode . "java")
                          (awk-mode . "awk")
                          (other . "linux")))
  ;; 'C-c .' to change style
#+end_src

4 commands
4.1 indentation commands
#+begin_src emacs-lisp
  (setq-default c-tab-always-indent t)
  ;; t: (the default), <TAB> simply indents the current line
  ;; nil: <TAB> (re)indents the line only if point is to the left of the first
  ;; non-whitespace character on the line. Otherwise it inserts some
  ;; whitespace (a tab or an equivalent number of spaces)
#+end_src

6 config basics
#+begin_src emacs-lisp
  (defun meu-cc-mode-common-hook ()
    ;; my customizations for all of c-mode and related modes
    (subword-mode 1) ; buffer local minor mode
    (c-toggle-electric-state 1)
    (c-toggle-auto-newline 1)
    (c-toggle-hungry-state 1)

    (setq fill-column 80)
    (which-function-mode 1))
  (add-hook 'c-mode-common-hook 'meu-cc-mode-common-hook)
  ;; you can also use 'subword-mode' in non-CC mode buffers by
  ;; typing M-x subword-mode
#+end_src
	use lambda functions

7

*** hooks
	c-initialization-hook
	c-mode-common-hook
		c-mode-hook
		c++-mode-hook
		objc-mode-hook
		java-mode-hook
		idl-mode-hook
		pike-mode-hook
		awk-mode-hook

*** 4.1 indentation commands

reindenting large sections of code
	<TAB>  'c-indent-command' indents the current line

-----

CC-mode specific:
C-M-q    'c-indent-exp'
C-c C-q  'c-indent-defun'
C-M-\    'indent-region' (the cc mode version)

C-M-h    'c-mark-function'
     While not strictly an indentation command, this is useful for
     marking the current top-level function or class definition as the
     current region.  As with ‘c-indent-defun’, this command operates on
     top-level constructs, and can’t be used to mark say, a Java method.

*** 4.2 comment commands

‘C-c C-c’ (‘comment-region’)
     This command comments out the lines that start in the region.  With
     a negative argument, it does the opposite: it deletes the comment
     delimiters from these lines.  *Note (emacs)Multi-Line Comments::,
     for fuller details.  ‘comment-region’ isn’t actually part of CC
     Mode; it is given a CC Mode binding for convenience.

M-;

*** 4.3 MOVEMENT COMMANDS

‘C-M-a’    (‘c-beginning-of-defun’) or ESC-C-a (my keyboard sucks!)
‘C-M-e’    (‘c-end-of-defun’)

M-a    c-beginning-of-statement
M-e    c-end-of-statement
	VERY USEFUL

‘C-c C-p’ (‘c-backward-conditional’)
‘C-c C-n’ (‘c-forward-conditional’)


M-x subword-backward
     These commands move backward or forward to the beginning of the
     next capitalized word.  With prefix argument N, move N times.  If N
     is negative, move in the opposite direction.


*SUBWORD-MODE*
	   TYPE ‘C-c C-w’ TO TOGGLE SUBWORD MODE ON AND OFF.  To make the mode
turn on automatically, put the following code in your ‘.emacs’:

     (add-hook 'c-mode-common-hook
               (lambda () (subword-mode 1)))

   As a bonus, you can also use ‘subword-mode’ in non-CC Mode buffers by
typing ‘M-x subword-mode’.

*** 4.4 filling and line breaking commands
M-q
M-j

*** 4.5 minor modes
/: line comment
\*: block comment
l: eletric mode
a: auto-newline mode
h: hungry delete
w: subword mode
	ex.:C/*lahw - comment style block and all the other mode enabled -

toggle commands
	‘C-c C-k’ (‘c-toggle-comment-style’)
	‘C-c C-l’ (‘c-toggle-electric-state’)  SICK!
	‘C-c C-a’ (‘c-toggle-auto-newline’)    SICK
	‘M-x c-toggle-hungry-state’            OK (good to delete tabs)
	‘M-x c-toggle-auto-hungry-state’
	‘C-c C-w’ (‘M-x subword-mode’)         SICK
	‘M-x c-toggle-syntactic-indentation’

*** 4.10 other commands
C-c :
C-c C-z

*** 5 font locking
5 font locking
5.2 faces
	NULL
5.3 documentation comments
	see M-x java-mode

The default value for ‘c-doc-comment-style’ is
‘((java-mode . javadoc) (pike-mode . autodoc) (c-mode . gtkdoc))’.

** eglot



** eshell
*** 2 entry points

eshell &optional arg
	C-N M-x eshell opens or switch to the (new) *eshell*<N>

-----

eshell-command

-----


** flymake

** cool commands
M-j

*M-S-x is SICK!!!*
	use it more often on a given mode, like CC MODE


* packages
** notes
M-x   list-packages <RET>

M-x package-install
M-x package-upgrade
M-x package-upgrade-all

M-x package-delete
M-x package-list-packages  ; to see the dependencies

   A package may “require” certain other packages to be installed,
because it relies on functionality provided by them.  When Emacs
installs such a package, it also automatically downloads and installs
any required package that is not already installed. [...]

** general
#+begin_src emacs-lisp
  ;;   By default, Emacs downloads packages from two archives: GNU ELPA
  ;;(https://elpa.gnu.org/) and NonGNU ELPA (https://elpa.nongnu.org/).
  (setq package-archives
        '(("gnu"   . "https://elpa.gnu.org/packages/")
          ("nongnu" . "https://elpa.nongnu.org/nongnu/")
          ("melpa" . "https://melpa.org/packages/")))
  ;; each list element must have the form ‘(ID . LOCATION)’

  ;; USING 'use-package' MACRO
  ;; ':ensure' makes use-package ask the Emacs package manager to install
  ;; a package if it is not already present on your system
  ;;
  ;; NOTE: you can customize the user option ‘use-package-always-ensure’
  ;; to a non-‘nil’ value if you want this behavior to be global for all
  ;; packages:
  (require 'use-package-ensure)
  (setq use-package-always-ensure t)
  ;;
  ;; ':defer' packages are only loaded when they are actually needed
  ;;
  ;; ':init' to evaluate Lisp _before_ a package is loaded
  ;; ':config' to execute code _after_ a package is loaded
#+end_src


** ORG-MODE

general
#+begin_src emacs-lisp
  ;; C-c C-x C-f 'org-emphasize'
  (setq org-hide-emphasis-markers t)

  ;;(setq org-pretty-entities t)
  ;;(setq org-use-sub-superscripts '{})

  (setq org-startup-with-inline-images t)
#+end_src

LATEX
#+begin_src emacs-lisp
  ;;(setq org-startup-with-latex-preview nil)
  ;;(setq org-preview-latex-default-process 'dvisvgm)
  (setq org-preview-latex-default-process 'dvipng)
  ;;
  (setq org-format-latex-options
        '(:scale 2.5
                 :html-scale 2.5
                 :foreground "White"
                 :background "Transparent"
                 :matchers ("$1" "$$" "\\(" "\\[")
  	       ))
  ;; 12.5.3 Using CDLaTeX to enter math (read Org Mode manual)
  ;; you'll need to install ‘cdlatex.el’ and ‘texmathp.el’
  ;; turn it on/off with 'M-x Org-cdlatex-mode'
  (use-package cdlatex
    :ensure t
    :hook (org-mode . org-cdlatex-mode)
    )
  ;; note: textmathp.el comes in AUCTeX
  (use-package tex
    :ensure auctex)
#+end_src

** pdf-tools
#+begin_src emacs-lisp
  (use-package pdf-tools
    :ensure t
    :config
    (pdf-tools-install) ; standard activation command
    ;;(pdf-loader-install) ; on demand loading, leads to faster startup time

    (setq-default pdf-view-display-size 'fit-page)

    (add-hook 'pdf-view-mode-hook
              (lambda ()
                (define-key pdf-view-mode-map (kbd "j") #'pdf-view-next-line-or-next-page)
                (define-key pdf-view-mode-map (kbd "k") #'pdf-view-previous-line-or-previous-page)
                (define-key pdf-view-mode-map (kbd "l") #'image-forward-hscroll)
                (define-key pdf-view-mode-map (kbd "h") #'image-backward-hscroll)
                (define-key pdf-view-mode-map (kbd "g") #'pdf-view-first-page)
                (define-key pdf-view-mode-map (kbd "G") #'pdf-view-last-page)))
    ;;:mode ("\\.pdf\\'" . pdf-view-mode) ; to enable
    :mode ("\\.null\\'" . pdf-view-mode) ; to disable
    )
  ;; CUSTOM CONFIGURATION FOR PDF COLOR (PDF-TOOLS)
  ;;(setq pdf-view-midnight-colors '("#ffffff" . "#000000"))
  (setq pdf-view-midnight-colors '("#f8f8f2" . "#282a36"))
  (add-hook 'pdf-view-mode-hook #'pdf-view-midnight-minor-mode)
  ;; it can be activated with C-c C-r m – or M-x pdf-view-midnight-minor-mode

  ;; FIX BLURRY PDFs
  (setq pdf-view-use-scaling t         ; for pdf-tools
        pdf-view-use-imagemagick nil)  ; for pdf-tools
  ;;(setq doc-view-resolution 100)     ; for docview
  ;; M-x set-variable doc-view-resolution
  ;; M-x doc-view-clear-cache
  ;; open the file again, and it should look much clearer
  ;; (though it'll take longer to load)
#+end_src

** windmove
#+begin_src emacs-lisp
  ;; enable windmove package
  ;;(windmove-default-keybindings) ; conflicts with org-mode
  ;; switch windows with S-<left>, S-<right>, S-<up>, S-<down>
#+end_src

* features
** scrolling

#+begin_src emacs-lisp
  ;;(setq scroll-preserve-screen-position t)
  ;;(setq scroll-conservatively 101)
  ;;(setq maximum-scroll-margin 0.5)
  ;;(setq scroll-margin 100)

;; I was using this
  ;;(setq scroll-margin 8
  ;;      scroll-conservatively 101
  ;;      scroll-preserve-screen-position t)
#+end_src

** autocomplete
#+begin_src emacs-lisp
  ;; * autocomplete
  ;; minibuffer selection (icomplete)
  (icomplete-mode t) ; default
  ;; M-x bs-show ; built into Emacs by default
  ;; M-x ibuffer ; built into Emacs by default
  ;; note: instead of C-x <LEFT> or C-x <RIGHT>, use C-x b <RET>
  ;;
  ;; option 1: minibuffer selection (ido)
  ;;(ido-mode 1)
  ;;
  ;; option 2: minibuffer selection (fido) (Fuzzy search)
  (fido-mode 1)
  ;;(fido-vertical-mode 1)
  (setq completion-styles '(flex))
  ;; fido is basically ido, but implemented the modern way
#+end_src

** search
#+begin_src emacs-lisp
  ;; slow?
  ;;(setq search-default-mode 'char-fold-to-regexp)
  ;;(setq char-fold-symmetric t)
  ;;	during search, you may use 'ss' instead of 'ß' ; slow?
  ;; note: apparently, character folding only applies to non-regexp
  ;;       searches (C-s)
  ;; note: to get folding-like behaviour in regex search: \(ss\|ß\)
#+end_src

** eshell
#+begin_src emacs-lisp
  ;;(global-font-lock-mode t)
  ;;
  ;; making passwords invisible in shell mode
  ;;(add-hook 'comint-output-filter-functions
  ;;'comint-watch-for-password-prompt)
#+end_src




* UNBINDING
#+begin_src emacs-lisp
  ;;(keymap-global-unset "C-x f")
#+end_src

* KEYBINDING
='C-c= followed by a =single character=
	strictly reserved for ~individuals own use~

-----

** text
*** fill-region
#+begin_src emacs-lisp
  ;; bind 'C-c e f' to 'fill-region  26.5 Code Alignment
  (keymap-global-set "C-c e f" 'fill-region)

  ;; NOTE: it uses the value of ' fill-column '
#+end_src

*** auto-fill-mode
#+begin_src emacs-lisp
  (keymap-global-set "C-c e a" 'auto-fill-mode)
#+end_src

** programming
*** code alignment
#+begin_src emacs-lisp

qtile/config.py
# ------------------------------------------------------------
# LIBRARIES
# ------------------------------------------------------------

from libqtile import bar, layout, qtile, widget
from libqtile.config import Click, Drag, Group, Key, Match, Screen
from libqtile.lazy import lazy
from libqtile.utils import guess_terminal

import os
import subprocess
from libqtile import hook


# ------------------------------------------------------------
# AUTOSTART
# ------------------------------------------------------------

@hook.subscribe.startup_once
def autostart():
    home = os.path.expanduser('~/.config/qtile/autostart.sh')
    subprocess.call(home)


# ------------------------------------------------------------
# GENERAL CONFIG
# ------------------------------------------------------------

mod = "mod4"
terminal = guess_terminal()
firefox = os.path.expanduser('/usr/bin/firefox'),


    # ********************************************************
    # DEFAULT KEYBINDS
    # --------------------------------------------------------

keys = [
    Key([mod], "h", lazy.layout.left()),
    Key([mod], "l", lazy.layout.right()),
    Key([mod], "j", lazy.layout.down()),
    Key([mod], "k", lazy.layout.up()),
    Key([mod], "space", lazy.layout.next(), desc="Move window focus to other window"),

    Key([mod, "shift"], "h", lazy.layout.shuffle_left()),
    Key([mod, "shift"], "l", lazy.layout.shuffle_right()),
    Key([mod, "shift"], "j", lazy.layout.shuffle_down()),
    Key([mod, "shift"], "k", lazy.layout.shuffle_up()),

    Key([mod, "control"], "h", lazy.layout.grow_left()),
    Key([mod, "control"], "l", lazy.layout.grow_right()),
    Key([mod, "control"], "j", lazy.layout.grow_down()),
    Key([mod, "control"], "k", lazy.layout.grow_up()),
    Key([mod], "n", lazy.layout.normalize(), desc="Reset all window sizes"),

    Key(
        [mod, "shift"],
        "Return",
        lazy.layout.toggle_split(),
        desc="Toggle between split and unsplit sides of stack",
    ),

    Key([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),
    Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    #Key([mod], "w", lazy.window.kill(), desc="Kill focused window"),
    Key(
        [mod],
        "f",
        lazy.window.toggle_fullscreen(),
        desc="Toggle fullscreen on the focused window",
    ),
    Key([mod], "t", lazy.window.toggle_floating(), desc="Toggle floating on the focused window"),
    Key([mod, "control"], "r", lazy.reload_config(), desc="Reload the config"),
    Key([mod, "control"], "q", lazy.shutdown(), desc="Shutdown Qtile"),
    #Key([mod], "r", lazy.spawncmd(), desc="Spawn a command using a prompt widget"),


    # ********************************************************
    # CUSTOM KEYBINDS
    # --------------------------------------------------------

    Key([mod, "shift", "control"], "h", lazy.layout.swap_column_left()),
    Key([mod, "shift", "control"], "l", lazy.layout.swap_column_right()),
    Key(["mod1", "shift"], "h", lazy.layout.swap_column_left()),
    Key(["mod1", "shift"], "l", lazy.layout.swap_column_right()),

    Key(
        [mod, "shift"],
        "w",
        lazy.layout.toggle_split(),
        desc="Toggle between split and unsplit sides of stack",
    ),

    Key([mod], "e", lazy.layout.next()),
    Key([mod], "q", lazy.next_layout()),
    Key([mod], "c", lazy.window.kill()),
    Key([mod], "r", lazy.window.toggle_floating()),
    Key([mod, "shift"], "r", lazy.reload_config()),
    Key([mod, "shift"], "e", lazy.shutdown()),
    Key([mod], "d", lazy.spawncmd()),

    # color inversion
    Key([mod], "i",
        lazy.spawn("/home/arch/.settings/color_inversion"),
    ),

    # pactl volume control
    Key([mod, "shift"], "x",
        lazy.spawn("/home/arch/.settings/inc_vol"),
    ),
    Key([mod, "shift"], "z",
        lazy.spawn("/home/arch/.settings/dec_vol"),
    ),

    # alacritty theme
    Key([mod, "control"], "p",
        lazy.spawn("/home/arch/.config/alacritty/theme_up.sh"),
    ),
    Key([mod, "control"], "n",
        lazy.spawn("/home/arch/.config/alacritty/theme_down.sh"),
    ),

    # flameshot
    Key([mod, "shift"], "f",
        lazy.spawn("flameshot gui"),
    ),
    Key(["mod1", "shift"], "s",
        lazy.spawn("flameshot gui"),
    ),
]

# ------------------------------------------------------------
# XMONAD SETTINGS
# ------------------------------------------------------------

#xmonad_keys = [
#    Key([mod], "h", lazy.layout.left()),
#    Key([mod], "l", lazy.layout.right()),
#    Key([mod], "j", lazy.layout.down()),
#    Key([mod], "k", lazy.layout.up()),
#    Key([mod, "shift"], "h", lazy.layout.swap_left()),
#    Key([mod, "shift"], "l", lazy.layout.swap_right()),
#    Key([mod, "shift"], "j", lazy.layout.shuffle_down()),
#    Key([mod, "shift"], "k", lazy.layout.shuffle_up()),
#    Key([mod, "shift"], "i", lazy.layout.grow()),
#    Key([mod], "m", lazy.layout.shrink()),
#    Key([mod], "n", lazy.layout.reset()),
#    Key([mod, "shift"], "n", lazy.layout.normalize()),
#    Key([mod], "o", lazy.layout.maximize()),
#    Key([mod, "shift"], "s", lazy.layout.toggle_auto_maximize()),
#    Key([mod, "shift"], "space", lazy.layout.flip()),
#]
#
#xmonad_layout = {
#"align":0,
#"auto_maximize":False,
#"border_focus":'#ff0000',
#"border_normal":'#000000',
#"border_width":2,
#"change_ratio":0.05,
#"change_size":20,
#"margin":0,
#"max_ratio":0.75,
#"min_ratio":0.25,
#"min_secondary_size":85,
#"new_client_position":'after_current',
#"ratio":0.5,
#"single_border_width":None,
#"single_margin":None,
#}

# ------------------------------------------------------------
# GROUPS
# ------------------------------------------------------------

# Add key bindings to switch VTs in Wayland.
# We can't check qtile.core.name in default config as it is loaded before qtile is started
# We therefore defer the check until the key binding is run by using .when(func=...)
for vt in range(1, 8):
    keys.append(
        Key(
            ["control", "mod1"],
            f"f{vt}",
            lazy.core.change_vt(vt).when(func=lambda: qtile.core.name == "wayland"),
            desc=f"Switch to VT{vt}",
        )
    )

groups = [Group(i) for i in "123456789" "10"]

for i in groups:
    keys.extend(
        [
            Key(
                [mod],
                i.name,
                lazy.group[i.name].toscreen(),
                desc=f"Switch to group {i.name}",
            ),
            Key(
                [mod, "shift"],
                i.name,
                lazy.window.togroup(i.name, switch_group=False),
                desc=f"move focused window to group {i.name}",
            ),
        ]
    )

# ------------------------------------------------------------
# LAYOUTS
# ------------------------------------------------------------

columns_layout = {
    "border_width": 2,
    "margin": 5,
    "border_focus": "DE1D95",
    "border_normal": "A662DF",
}

floating_layout = {
    "border_focus":"ffffff",
    "border_normal":"000000",
    "border_width": 2,
    "fullscreen_border_width":0,
    "max_border_width":0,
}

GroupBox_theme0 = {

    "active":'2DDE1D',
    "background":'00000080',
    "block_highlight_text_color":None,
    "borderwidth":1,
    "center_aligned":True,
    "disable_drag":True,
    "fmt":'{}',
    "font":'sans',
    "fontshadow":None,
    "fontsize":20,
    "foreground":'ffffff',
    "hide_crash":False,
    "hide_unused":False,
    "highlight_method":'text',
    "inactive":'A662DF',
    "invert_mouse_wheel":True,
    #"margin":3,
    "margin_x":10,
    "margin_y":3,
    "markup":False,
    "max_chars":0,
    "mouse_callbacks":{},
    "other_current_screen_border":'F2211D',
    "other_screen_border":'1DE4F2',
    #padding:10,
    #padding_x:None,
    #padding_y:None,
    "rotate":True,
    "rounded":True,
    #spacing:-10,
    "this_current_screen_border":'DE1D95',
    "this_screen_border":'F21DEF',
    "toggle":True,
}

layouts = [
     layout.Columns(**columns_layout),
     layout.MonadTall(),
     layout.Max(),

    # layout.MonadWide(),
    # layout.Bsp(),

    # strange
    # layout.Stack(num_stacks=2),
    # layout.Matrix(),
    # layout.Tile(),
    # layout.RatioTile(),
    # layout.TreeTab(), # really strange
    # layout.VerticalTile(),
    # layout.Zoomy(), # really strange
]

widget_defaults = dict(
    font="sans",
    fontsize=20,
    padding=4,
)
extension_defaults = widget_defaults.copy()

screens = [
    Screen(
        bottom=bar.Gap(5),
        left=bar.Gap(5),
        right=bar.Gap(5),

        top=bar.Bar(
            [
                widget.GroupBox(**GroupBox_theme0),
                widget.CurrentLayout(),
                widget.Prompt(),
                widget.WindowName(),
                widget.Systray(),
                widget.Clock(format="%a  %d/%m/%Y  %I:%M %p"),
                #widget.QuickExit(),
            ],
            background='#00000080',
            #background='#FFFFFF80',
            size=35,
            #border_width=[2, 1, 2, 1],
            #border_color=["ff00ff80", "00000080", "ff00ff80", "00000080"]
        ),
        # You can uncomment this variable if you see that on X11 floating resize/moving is laggy
        # x11_drag_polling_rate = 60,
    ),
]

# Drag floating layouts.
mouse = [
    Drag([mod], "Button1", lazy.window.set_position_floating(), start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(), start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front()),
]

dgroups_key_binder = None
dgroups_app_rules = []  # type: list
follow_mouse_focus = True
bring_front_click = False
floats_kept_above = True
cursor_warp = False
floating_layout = layout.Floating(
    float_rules=[
        *layout.Floating.default_float_rules,
        Match(wm_class="confirmreset"),  # gitk
        Match(wm_class="makebranch"),  # gitk
        Match(wm_class="maketag"),  # gitk
        Match(wm_class="ssh-askpass"),  # ssh-askpass
        Match(title="branchdialog"),  # gitk
        Match(title="pinentry"),  # GPG key password entry

        Match(wm_class="PacketTracer"),  # GPG key password entry
        Match(wm_class="discord"),  # GPG key password entry
    ]
)
auto_fullscreen = True
focus_on_window_activation = "smart"
reconfigure_screens = True

# If things like steam games want to auto-minimize themselves when losing
# focus, should we respect this or not?
auto_minimize = True

# When using the Wayland backend, this can be used to configure input devices.
wl_input_rules = None

# xcursor theme (string or None) and size (integer) for Wayland backend
wl_xcursor_theme = None
wl_xcursor_size = 24

# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits; you can see several discussions on the
# mailing lists, GitHub issues, and other WM documentation that suggest setting
# this string if your java app doesn't work correctly. We may as well just lie
# and say that we're a working one by default.
#
# We choose LG3D to maximize irony: it is a 3D non-reparenting WM written in
# java that happens to be on java's whitelist.
wmname = "LG3D"
